File: 19_remove_nth_node_from_end_of_list/pramoda/solution.c
#include <stdio.h>
#include <stdlib.h>

// Definition for singly-linked list.
struct ListNode {
    int val;
    struct ListNode *next;
};

void node(struct ListNode* head) {
    while (head != NULL) {
        printf("%d ", head->val);
        head = head->next;
    }
}

struct ListNode* removeNthFromEnd(struct ListNode* head, int n) {
    struct ListNode dummy = {0, head};

    struct ListNode *fast = &dummy;
    struct ListNode *slow = &dummy;

    for (int i = 0; i < n; i++) {
        fast = fast->next;
    }

    while (fast->next) {
        fast = fast->next;
        slow = slow->next;
    }

    slow->next = slow->next->next;

    return dummy.next;
}

int main() {
    struct ListNode *nodes = malloc(5 * sizeof(struct ListNode));

    for (int i = 0; i < 5; i++) {
        nodes[i].val = i + 1;
        nodes[i].next = (i < 4) ? &nodes[i + 1] : NULL;
    }

    struct ListNode *head = &nodes[0];
    
    printf("Before removal:\n");
    node(head);
    printf("\n");
    
    head = removeNthFromEnd(head, 2);
    
    printf("After removal:\n");
    node(head);
    printf("\n");

    free(nodes);

    return 0;
}

File: 19_remove_nth_node_from_end_of_list/pramoda/solution.py
from typing import Optional

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

def node(head: ListNode):
    while head:
        print(head.val, "->", end=" ")
        head = head.next

class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0, head)
        fast = slow = dummy

        for _ in range(n):
            fast = fast.next

        while fast.next:
            fast = fast.next
            slow = slow.next

        slow.next = slow.next.next

        return dummy.next

s = Solution()
l = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(5)))))
node(s.removeNthFromEnd(l, 2))
